signer/core/signed_data_test.go

```diff
@@ -0,0 +1,715 @@
+	}
+	if err != ErrRequestDenied {
+		t.Errorf("Expected ErrRequestDenied! %v", err)
+	}
+	// text/plain
+	control <- "Y"
+	control <- "a_long_password"
+	signature, err = api.SignData(context.Background(), TextPlain.Mime, a, hexutil.Encode([]byte("EHLO world")))
+	if err != nil {
+		t.Fatal(err)
+	}
+	if signature == nil || len(signature) != 65 {
+		t.Errorf("Expected 65 byte signature (got %d bytes)", len(signature))
+	}
+	// data/typed
+	control <- "Y"
+	control <- "a_long_password"
+	signature, err = api.SignTypedData(context.Background(), a, typedData)
+	if err != nil {
+		t.Fatal(err)
+	}
+	if signature == nil || len(signature) != 65 {
+		t.Errorf("Expected 65 byte signature (got %d bytes)", len(signature))
+	}
+}
+
+func TestHashStruct(t *testing.T) {
+	hash, err := typedData.HashStruct(typedData.PrimaryType, typedData.Message)
+	if err != nil {
+		t.Fatal(err)
+	}
+	mainHash := fmt.Sprintf("0x%s", common.Bytes2Hex(hash))
+	if mainHash != "0xc52c0ee5d84264471806290a3f2c4cecfc5490626bf912d01f240d7a274b371e" {
+		t.Errorf("Expected different hashStruct result (got %s)", mainHash)
+	}
+
+	hash, err = typedData.HashStruct("EIP712Domain", typedData.Domain.Map())
+	if err != nil {
+		t.Error(err)
+	}
+	domainHash := fmt.Sprintf("0x%s", common.Bytes2Hex(hash))
+	if domainHash != "0xf2cee375fa42b42143804025fc449deafd50cc031ca257e0b194a650a912090f" {
+		t.Errorf("Expected different domain hashStruct result (got %s)", domainHash)
+	}
+}
+
+func TestEncodeType(t *testing.T) {
+	domainTypeEncoding := string(typedData.EncodeType("EIP712Domain"))
+	if domainTypeEncoding != "EIP712Domain(string name,string version,uint256 chainId,address verifyingContract)" {
+		t.Errorf("Expected different encodeType result (got %s)", domainTypeEncoding)
+	}
+
+	mailTypeEncoding := string(typedData.EncodeType(typedData.PrimaryType))
+	if mailTypeEncoding != "Mail(Person from,Person to,string contents)Person(string name,address wallet)" {
+		t.Errorf("Expected different encodeType result (got %s)", mailTypeEncoding)
+	}
+}
+
+func TestTypeHash(t *testing.T) {
+	mailTypeHash := fmt.Sprintf("0x%s", common.Bytes2Hex(typedData.TypeHash(typedData.PrimaryType)))
+	if mailTypeHash != "0xa0cedeb2dc280ba39b857546d74f5549c3a1d7bdc2dd96bf881f76108e23dac2" {
+		t.Errorf("Expected different typeHash result (got %s)", mailTypeHash)
+	}
+}
+
+func TestEncodeData(t *testing.T) {
+	hash, err := typedData.EncodeData(typedData.PrimaryType, typedData.Message, 0)
+	if err != nil {
+		t.Fatal(err)
+	}
+	dataEncoding := fmt.Sprintf("0x%s", common.Bytes2Hex(hash))
+	if dataEncoding != "0xa0cedeb2dc280ba39b857546d74f5549c3a1d7bdc2dd96bf881f76108e23dac2fc71e5fa27ff56c350aa531bc129ebdf613b772b6604664f5d8dbe21b85eb0c8cd54f074a4af31b4411ff6a60c9719dbd559c221c8ac3492d9d872b041d703d1b5aadf3154a261abdd9086fc627b61efca26ae5702701d05cd2305f7c52a2fc8" {
+		t.Errorf("Expected different encodeData result (got %s)", dataEncoding)
+	}
+}
+
+func TestMalformedData1(t *testing.T) {
```

Comment:
please describe in what respect the data is malformed, so make it more explicit to the person for whom it will fail